### PR TITLE
Use LDC for cross-compilation Win{32,64} binaries

### PR DESCRIPTION
**Pull request by wilzbach**
Follow-up to https://github.com/dlang-community/D-Scanner/pull/623

Based on @BBasile's comment: https://github.com/dlang-community/D-Scanner/pull/623#issue-184847942

### CODE REVIEW

**Comment by BBasile** [comment]
As said in the other PR, there's no need to use the DFLAG as i did when experimenting. Add this to the makefile

```make
ldcxwin64:
	${LDC} ${LDC_FLAGS} -of=bin/dscanner ${VERSIONS} ${INCLUDE_PATHS} ${SRC} -Jbin -mtriple=x86_64-windows-msvc
```

And then oviously just `make ldcxwin64` is enough to cross compile.

**Comment by wilzbach** [comment]
This PR will now use LDC for cross-compilation for both x86 and x86_64 and produce these binaries:

[dscanner-v0.5.7-windows-x86_64.zip](https://github.com/dlang-community/D-Scanner/files/2728013/dscanner-v0.5.7-windows-x86_64.zip)
[dscanner-v0.5.7-windows-x86.zip](https://github.com/dlang-community/D-Scanner/files/2728014/dscanner-v0.5.7-windows-x86.zip)

It would be great to get feedback from a Windows user on these binaries.

**Comment by wilzbach** [comment]
> As said in the other PR, there's no need to use the DFLAG as i did when experimenting. Add this to the makefile
> 
> ```makefile
> ldcxwin64:
> 	${LDC} ${LDC_FLAGS} -of=bin/dscanner ${VERSIONS} ${INCLUDE_PATHS} ${SRC} -Jbin -mtriple=x86_64-windows-msvc
> ```
> 
> And then oviously just `make ldcxwin64` is enough to cross compile.

Let's do beautification at the end (i.e. once Travis is green) and in this case actually using `DFLAGS` isn't too bad because then I can use LDC for cross-compilation to 32-bit too.

**Comment by wilzbach** [comment]
Yup, this is also working on Travis :tada: 

image

**Comment by BBasile** [comment]
I'll go on win to test them...

**Comment by wilzbach** [comment]
Thanks!
I guess now the only remaining bit would be to ensure that all LDC release flags are used for the binaries (the current Makefiles setup doesn't use any).

**Comment by wilzbach** [comment]
So apparently there was some old binaries in my `bin` directory and it could be possible that the script packaged those for the zip archive.
Would you mind verifying these binaries again:

[dscanner-v0.6.0-windows-x86_64.zip](https://github.com/dlang-community/D-Scanner/files/2728258/dscanner-v0.6.0-windows-x86_64.zip)
[dscanner-v0.6.0-windows-x86.zip](https://github.com/dlang-community/D-Scanner/files/2728259/dscanner-v0.6.0-windows-x86.zip)

(this now also includes LDC with all optimization levels except for full LTO as this is probably a lot harder to get to work with cross-compilation).

**Comment by wilzbach** [comment]
> I let you merge when you're happy with the PR.

Yup, I am happy now. We just need to make once sure that all these binaries are properly built.

> Then DCD and DFMT will need this change too.

I split up the LDC setup in a separate file, s.t. it's easier to copy the file over (though on the long run we might want to add stuff like this to a shared `tools` repo ...)
Anyhow:

https://github.com/dlang-community/DCD/pull/568
https://github.com/dlang-community/dfmt/pull/420

**Comment by wilzbach** [comment]
Hmm so I was testing this in one of the [free VirtualBox for Windows IE](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/), but it complains about the missing `vcruntime140.dll` here:

image

And wine just aborts with "wine: Bad EXE format for ". So I guess we must have missed sth. here.

**Comment by BBasile** [comment]
> So apparently there was some old binaries

Yes i should have seen that 0.5.7 was not the latest version

> We just need to make once sure that all these binaries are properly built.

They work here. The Only issue i see is that there a missing switch (-StdDisableLogging) because warns of type "Could not locate object.d or object.di ..." are visible while they should not

**Comment by wilzbach** [comment]
> Actually the dependence on the ms C library is specified

Hmm fair enough, but shouldn't there be a way to statically bundle it like we can do for Linux? : thinking:

**Comment by BBasile** [comment]
> shouldn't there be a way to statically bundle

The version produced by dmd with `-m64` also has a dependence on a dynamic library (msvcr100.dll) so I'd say that either it's not possible or there are licensing issue. Maybe one of the LDC developer will know better.

File: release-windows.sh [hunk 1]
```diff
@@ -1,26 +1,21 @@
...
 
-# Step 2: Run DMD via wineconsole
-archiveName="dscanner-$VERSION-$OS-$ARCH_SUFFIX.zip"
+# Run LDC with cross-compilation
+archiveName="$BIN_NAME-$VERSION-$OS-$ARCH_SUFFIX.zip"
 echo "Building $archiveName"
 mkdir -p bin
-git describe --tags > bin/githash.txt # no git installed under Wine
-DC="$DIR/dmd2/windows/bin/dmd.exe" wine cmd /C build.bat
+DC=ldmd2 make ldcbuild
```
Comment (BBasile) [inline release-windows.sh]: DC=ldmd2 is useless since you `make ldcbuild`.

File: makefile [hunk 2]
```diff
@@ -26,6 +26,9 @@ INCLUDE_PATHS = \
 VERSIONS =
 DEBUG_VERSIONS = -version=dparse_verbose
 DMD_FLAGS = -w -release -O -Jbin -od${OBJ_DIR} -version=StdLoggerDisableWarning
+override DMD_FLAGS += $(DFLAGS)
+override LDC_FLAGS += $(DFLAGS)
+override GDC_FLAGS += $(DFLAGS)
 DMD_TEST_FLAGS = -w -g -Jbin -version=StdLoggerDisableWarning
 override LDC_FLAGS += -O5 -release -oq
```
Comment (BBasile) [inline makefile]: here must be added -version=StdLoggerDisableWarning. for FDC flags too.

**Comment by LaurentTreguier** [comment]
> Hmm so I was testing this in one of the [free VirtualBox for Windows IE](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/), but it complains about the missing `vcruntime140.dll` here:
> 
> image

I tried compiling a binary with LDC and testing it on the Windows 7 image; I got the `missing vcruntime140.dll` error when LDC uses LLD, but not when using Microsoft's linker.
The Windows LDC archive has a [`msvcEnv.bat`](https://github.com/ldc-developers/ldc/blob/master/vcbuild/msvcEnv.bat) script that can be used to load the necessary environment variables (`msvcEnv x64` or `msvcEnv x86`).

> And wine just aborts with "wine: Bad EXE format for ". So I guess we must have missed sth. here.

I got this when trying to launch the x86 binary with wine64, but not with the x64 binary.

**Comment by BBasile** [comment]
@wilzbach  can you address the two small points i've raised after testing the right binaries ?
The dll is not an issue. People will have to install the MS installer for this lib, that's not a problem.
It's very likely that on a real system it's already there (i.e a system used daily since months or years.

So it's almost good now, i don't understand why you stop two meters before the finish line.

**Pull request merged**